src/utilities/modules/vaccines.js

```diff
@@ -0,0 +1,152 @@
+/**
+ * mSupply Mobile
+ * Sustainable Solutions (NZ) Ltd. 2019
+ */
+
+/**
+ * Utility and helper methods for the vaccine
+ * module.
+ */
+
+/**
+ * Extracts breaches from a set of sensor logs.
+ * Breach: Sequential sensorLog objects whose temperature is outside
+ * the lower or uppound limits, form a breach. Each array returned
+ * contains all sensorLogs which form a breach INCLUDING the sensorLog
+ * prior to the breach, and after the breach, if they exist. There may
+ * be duplicate non-breach sensorLogs as breach delimiters if there is
+ * only one sensorLog between breaches.
+ *
+ * Method sorts the sensorLogs by timestamp, iterates and pushes onto a
+ * stack. Only one unBreached sensorLog is kept at the tail of the stack.
+ * All breached sensorLogs are pushed, until a non breached log, where this
+ * marks a breach has been found. Store these sensorLogs and reset the stack.
+ *
+ * @param  {Realm.results} sensorLogs array of sensorLog objects
+ * @return {Array<Realm.results<SensorLog>} A 2D array of Realm.results. Example:
+ * [ [ sensorLog1, sensorLog2 .. ], [ sensorLog1, sensorLog2, sensorLog3, .. ]]
+ * where each element is a Realm.results<SensorLog>
+ */
+export const extractBreaches = ({ sensorLogs = [], database }) => {
+  if (!(database && sensorLogs.length > 0)) return [];
+
+  const sensorLogStack = [];
+  const breaches = [];
+
+  // Sort the sensor logs by timestamp.
+  sensorLogs.sorted('timestamp', true);
+
+  sensorLogs.forEach(sensorLog => {
+    const { isInBreach } = sensorLog;
+    const { length: stackLength } = sensorLogStack;
+    // If the stack is empty, just push this sensorLog.
+    if (stackLength === 0) sensorLogStack.push(sensorLog);
+    // Any breached sensorLogs are always pushed onto the stack.
+    else if (isInBreach) sensorLogStack.push(sensorLog);
+    // If both head of the stack and this sensorLog are unbreached,
+    // replace the head. This can occurs when the length is 1. Two
+    // sequential unbreached logs are never pushed.
+    else if (!sensorLogStack[stackLength - 1].isInBreach && !isInBreach) {
+      sensorLogStack.pop();
+      sensorLogStack.push(sensorLog);
+      // In all other cases, a breach has been found. Push the last sensorlog,
+      // store the breach and reset the stack. Push this sensorLog back
+      // onto the stack as the new potential delimiter for the next breach.
+    } else {
+      sensorLogStack.push(sensorLog);
+      breaches.push([...sensorLogStack]);
+      sensorLogStack.length = 0;
```

Comment:
that's very cool, but can you please `sensorLogStack = []`, i just not sure if .length = 0 will aways be `supported`

Reply:
c43838fcaeb90c506c44543a599b7c4b38fafb47